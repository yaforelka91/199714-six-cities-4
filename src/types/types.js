import PropTypes from 'prop-types';
import {CardView} from '../const';

export const offerShape = PropTypes.shape({
  city: PropTypes.shape({
    name: PropTypes.string.isRequired,
    coords: PropTypes.arrayOf(PropTypes.number).isRequired,
    zoom: PropTypes.number.isRequired,
  }).isRequired,
  id: PropTypes.number.isRequired,
  coords: PropTypes.arrayOf(PropTypes.number).isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.arrayOf(PropTypes.string).isRequired,
  pictures: PropTypes.arrayOf(PropTypes.string).isRequired,
  price: PropTypes.number.isRequired,
  type: PropTypes.string.isRequired,
  isPremium: PropTypes.bool.isRequired,
  isFavorite: PropTypes.bool.isRequired,
  rating: PropTypes.number.isRequired,
  bedrooms: PropTypes.number.isRequired,
  guests: PropTypes.number.isRequired,
  services: PropTypes.arrayOf(PropTypes.string).isRequired,
  host: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
    isSuper: PropTypes.bool.isRequired,
  }),
});

export const cityShape = PropTypes.shape({
  name: PropTypes.string.isRequired,
  coords: PropTypes.arrayOf(PropTypes.number).isRequired,
  zoom: PropTypes.number.isRequired,
});

export const reviewShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
    picture: PropTypes.string.isRequired,
  }).isRequired,
  rating: PropTypes.number.isRequired,
  feedback: PropTypes.string.isRequired,
  visitTime: PropTypes.string.isRequired,
});

export const appTypes = {
  activeCard: PropTypes.oneOfType([offerShape, PropTypes.bool]).isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  onOfferTitleClick: PropTypes.func.isRequired,
};

export const buttonTypes = {
  className: PropTypes.string.isRequired,
  isDisabled: PropTypes.bool.isRequired,
  onButtonClick: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

export const citiesTypes = {
  offers: PropTypes.arrayOf(offerShape).isRequired,
  activeCity: PropTypes.string.isRequired,
  activeItem: PropTypes.number.isRequired,
  onActiveChange: PropTypes.func.isRequired,
};

export const cityListTypes = {
  citiesList: PropTypes.arrayOf(cityShape).isRequired,
  activeCity: cityShape.isRequired,
  onCityNameClick: PropTypes.func.isRequired,
};

export const cityListItemTypes = {
  city: PropTypes.string.isRequired,
  classNameLink: PropTypes.string.isRequired,
  onCityNameClick: PropTypes.func.isRequired,
};

export const headerTypes = {
  authorizationStatus: PropTypes.string.isRequired,
  isMain: PropTypes.bool.isRequired,
  userData: PropTypes.shape({
    email: PropTypes.string,
    picture: PropTypes.string,
  }).isRequired,
};

export const loginTypes = {
  activeCity: PropTypes.string.isRequired,
  validationError: PropTypes.string.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
};

export const mainTypes = {
  offersList: PropTypes.arrayOf(offerShape).isRequired,
  activeCity: PropTypes.string.isRequired,
  citiesList: PropTypes.arrayOf(PropTypes.string).isRequired,
  errorType: PropTypes.string.isRequired,
  onCityNameClick: PropTypes.func.isRequired,
};

export const mapTypes = {
  offers: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        coords: PropTypes.arrayOf(PropTypes.number).isRequired,
      })
  ).isRequired,
  activeCard: PropTypes.number.isRequired,
  className: PropTypes.string,
};

export const noPlacesTypes = {
  city: PropTypes.string.isRequired,
};

export const offerCardTypes = {
  offer: offerShape.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
  classNameForImage: PropTypes.string.isRequired,
  classNameForInfo: PropTypes.string.isRequired,
  onOfferTitleClick: PropTypes.func.isRequired,
  onOfferCardEnter: PropTypes.func.isRequired,
  onFavoriteButtonClick: PropTypes.func.isRequired,
};

export const offerListTypes = {
  offers: PropTypes.arrayOf(offerShape).isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
  viewMode: PropTypes.oneOf([CardView.CITIES, CardView.NEAR, CardView.FAVORITES]).isRequired,
  onOfferTitleClick: PropTypes.func.isRequired,
  onOfferCardEnter: PropTypes.func.isRequired,
  onFavoriteButtonClick: PropTypes.func.isRequired,
};

export const offerPageTypes = {
  offer: offerShape.isRequired,
  offersList: PropTypes.arrayOf(offerShape).isRequired,
  onOfferTitleClick: PropTypes.func.isRequired,
  onFavoriteButtonClick: PropTypes.func.isRequired,
};

export const pageTypes = {
  className: PropTypes.string.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  userData: PropTypes.shape({
    email: PropTypes.string,
    picture: PropTypes.string,
  }).isRequired,
  isLoading: PropTypes.bool.isRequired,
  isMain: PropTypes.bool.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

export const placesTypes = {
  offers: PropTypes.arrayOf(offerShape).isRequired,
  city: PropTypes.string.isRequired,
  renderSorting: PropTypes.func.isRequired,
  className: PropTypes.string,
  onOfferCardEnter: PropTypes.func.isRequired,
};

export const ratingTypes = {
  selectedRating: PropTypes.string.isRequired,
  onRatingChange: PropTypes.func.isRequired,
};

export const reviewCardTypes = {
  review: reviewShape.isRequired,
};

export const reviewFormTypes = {
  offerId: PropTypes.number.isRequired,
  serverError: PropTypes.string.isRequired,
  isValid: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  renderRating: PropTypes.func.isRequired,
  renderTextarea: PropTypes.func.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
};

export const reviewsTypes = {
  reviews: PropTypes.arrayOf(reviewShape).isRequired,
  reviewsCount: PropTypes.number.isRequired,
  authorizationStatus: PropTypes.string.isRequired,
  offerId: PropTypes.number.isRequired,
  className: PropTypes.string.isRequired,
  onReviewFormSubmit: PropTypes.func.isRequired,
};

export const sortingTypes = {
  activeSorting: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onSortItemClick: PropTypes.func.isRequired,
  onToggleMenu: PropTypes.func.isRequired,
  onSelectMenu: PropTypes.func.isRequired,
};

export const tabsTypes = {
  activeItem: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.string).isRequired,
  onActiveChange: PropTypes.func.isRequired,
  renderItem: PropTypes.func.isRequired,
  classNameForList: PropTypes.string.isRequired,
};

export const textareaTypes = {
  currentMessage: PropTypes.string.isRequired,
  maxLength: PropTypes.number.isRequired,
  onTextareaChange: PropTypes.func.isRequired,
};
