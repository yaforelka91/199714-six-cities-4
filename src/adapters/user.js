const adaptUser = (user) => {
  return {
    id: user.id,
    email: user.email,
    name: user.name,
    picture: user.avatar_url,
    isSuper: user.is_pro,
  };
};

export default adaptUser;
