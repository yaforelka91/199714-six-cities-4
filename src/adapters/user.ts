import User from '../interfaces/user';

const adaptUser = (user): User => ({
  id: user.id,
  email: user.email,
  name: user.name,
  picture: user.avatar_url,
  isSuper: user.is_pro,
});

export default adaptUser;
