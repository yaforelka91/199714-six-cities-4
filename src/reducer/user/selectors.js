import NameSpace from '../name-space';

const NAME_SPACE = NameSpace.USER;

export const getAuthorizationStatus = (state) => state[NAME_SPACE].authorizationStatus;

export const getAuthorizationProgress = (state) => state[NAME_SPACE].isAuthorizationInProgress;

export const getUserData = (state) => state[NAME_SPACE].userData;

export const getError = (state) => state[NAME_SPACE].errorType;
