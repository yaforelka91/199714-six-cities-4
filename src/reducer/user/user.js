import {extend} from '../../utils.js';
import adaptUser from '../../adapters/user.js';
import adaptError from '../../adapters/error.js';
import {Operation as DataOperation} from '../data/data.js';
import {Error} from '../../api.js';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  userData: {},
  isAuthorizationInProgress: true,
  errorType: ``,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_USER_DATA: `SET_USER_DATA`,
  CHANGE_PROGRESS_STATUS: `CHANGE_PROGRESS_STATUS`,
  CATCH_ERROR: `CATCH_ERROR`,
};

const ActionCreator = {
  requireAuthorization: (status) => {
    return {
      type: ActionType.REQUIRED_AUTHORIZATION,
      payload: status,
    };
  },
  setUserData: (authData) => {
    return {
      type: ActionType.SET_USER_DATA,
      payload: authData,
    };
  },
  changeProgressStatus: (isLoaded) => {
    return {
      type: ActionType.CHANGE_PROGRESS_STATUS,
      payload: isLoaded,
    };
  },
  catchError: (errorMessage) => {
    return {
      type: ActionType.CATCH_ERROR,
      payload: errorMessage,
    };
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_USER_DATA:
      return extend(state, {
        userData: action.payload,
      });

    case ActionType.CHANGE_PROGRESS_STATUS:
      return extend(state, {
        isAuthorizationInProgress: action.payload,
      });

    case ActionType.CATCH_ERROR:
      return extend(state, {
        errorType: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setUserData(adaptUser(response.data)));
      })
      .catch((err) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        throw err;
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setUserData(adaptUser(response.data)));
        dispatch(ActionCreator.catchError(``));
        dispatch(DataOperation.loadOffers());
      })
      .catch((err) => {
        dispatch(ActionCreator.changeProgressStatus(false));

        if (err.response.status === Error.BAD_REQUEST) {
          dispatch(ActionCreator.catchError(adaptError(err.response.data.error)));
          return;
        }

        throw err;
      });
  },
};

export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
