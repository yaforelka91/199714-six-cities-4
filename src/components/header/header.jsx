import React from 'react';
import {AuthorizationStatus} from '../../reducer/user/user';
import {headerTypes} from '../../types/types.js';

const Header = ({authorizationStatus, userData}) => {
  return (
    <header className="header">
      <div className="container">
        <div className="header__wrapper">
          <div className="header__left">
            <a className="header__logo-link header__logo-link--active">
              <img className="header__logo" src="img/logo.svg" alt="6 cities logo" width="81" height="41" />
            </a>
          </div>
          <nav className="header__nav">
            <ul className="header__nav-list">
              <li className="header__nav-item user">
                <a className="header__nav-link header__nav-link--profile" href="#">
                  <div
                    className="header__avatar-wrapper user__avatar-wrapper"
                  >
                    {authorizationStatus === AuthorizationStatus.AUTH && <img src={`https://htmlacademy-react-3.appspot.com/six-cities${userData.picture}`} alt="User avatar" />}
                  </div>
                  {authorizationStatus === AuthorizationStatus.AUTH && <span className="header__user-name user__name">{userData.email}</span>}
                  {authorizationStatus === AuthorizationStatus.NO_AUTH && <span className="header__login">Sign in</span>}
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
  );
};

Header.propTypes = headerTypes;

export default Header;
