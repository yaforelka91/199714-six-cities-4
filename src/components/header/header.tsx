import React from 'react';
import {Link, useLocation} from 'react-router-dom';
import {AuthorizationStatus} from '../../reducer/user/user';
import {AppRoute} from '../../types';
import {connect} from 'react-redux';
import {getAuthorizationStatus, getUserData} from '../../reducer/user/selectors';
import User from '../../interfaces/user';

type Props = {
  authorizationStatus: string;
  userData: User;
};

const Header: React.FC<Props> = (props: Props) => {
  const {authorizationStatus, userData} = props;
  return (
    <header className="header">
      <div className="container">
        <div className="header__wrapper">
          <div className="header__left">
            {useLocation().pathname === AppRoute.ROOT ?
              <span className="header__logo-link header__logo-link--active">
                <img className="header__logo" src="/img/logo.svg" alt="6 cities logo" width="81" height="41" />
              </span> :
              <Link className="header__logo-link" to={AppRoute.ROOT}>
                <img className="header__logo" src="/img/logo.svg" alt="6 cities logo" width="81" height="41" />
              </Link>
            }
          </div>
          <nav className="header__nav">
            <ul className="header__nav-list">
              <li className="header__nav-item user">
                <Link
                  className="header__nav-link header__nav-link--profile"
                  to={authorizationStatus === AuthorizationStatus.NO_AUTH ? AppRoute.LOGIN : AppRoute.FAVORITES}
                >
                  <div
                    className="header__avatar-wrapper user__avatar-wrapper"
                  >
                    {authorizationStatus === AuthorizationStatus.AUTH && <img className='user__avatar' src={`https://htmlacademy-react-3.appspot.com/six-cities${userData.picture}`} alt="User avatar" />}
                  </div>
                  {authorizationStatus === AuthorizationStatus.AUTH && <span className="header__user-name user__name">{userData.email}</span>}
                  {authorizationStatus === AuthorizationStatus.NO_AUTH && <span className="header__login">Sign in</span>}
                </Link>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
  );
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  userData: getUserData(state),
});

export {Header};
export default connect(mapStateToProps)(Header);
