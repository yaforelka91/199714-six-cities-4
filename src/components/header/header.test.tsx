import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Header} from './header';
import {AuthorizationStatus} from '../../reducer/user/user';
import {MemoryRouter, Router, Route} from 'react-router-dom';
import history from '../../history';
import {AppRoute} from '../../types';


describe(`HeaderSnapshot`, () => {
  it(`should render Header for guest`, () => {
    const tree = renderer.create(
        <Router
          history={history}
        >
          <Header
            authorizationStatus={AuthorizationStatus.NO_AUTH}
            userData={null}
          />
        </Router>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render Header for user`, () => {
    const tree = renderer.create(
        <Router
          history={history}
        >
          <Header
            authorizationStatus={AuthorizationStatus.AUTH}
            userData={{
              id: 1,
              name: `user`,
              email: `test@test.com`,
              picture: `/pic.jpg`,
              isSuper: true,
            }}
          />
        </Router>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render Header with Link`, () => {
    const tree = renderer.create(
        <MemoryRouter initialEntries={[`${AppRoute.LOGIN}`]}>
          <Route
            path={AppRoute.LOGIN}
          >
            <Header
              authorizationStatus={AuthorizationStatus.AUTH}
              userData={{
                id: 1,
                name: `user`,
                email: `test@test.com`,
                picture: `/pic.jpg`,
                isSuper: true,
              }}
            />
          </Route>
        </MemoryRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
