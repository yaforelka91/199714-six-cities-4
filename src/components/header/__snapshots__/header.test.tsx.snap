// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderSnapshot should render Header for guest 1`] = `
<header
  className="header"
>
  <div
    className="container"
  >
    <div
      className="header__wrapper"
    >
      <div
        className="header__left"
      >
        <span
          className="header__logo-link header__logo-link--active"
        >
          <img
            alt="6 cities logo"
            className="header__logo"
            height="41"
            src="/img/logo.svg"
            width="81"
          />
        </span>
      </div>
      <nav
        className="header__nav"
      >
        <ul
          className="header__nav-list"
        >
          <li
            className="header__nav-item user"
          >
            <a
              className="header__nav-link header__nav-link--profile"
              href="/login"
              onClick={[Function]}
            >
              <div
                className="header__avatar-wrapper user__avatar-wrapper"
              />
              <span
                className="header__login"
              >
                Sign in
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</header>
`;

exports[`HeaderSnapshot should render Header for user 1`] = `
<header
  className="header"
>
  <div
    className="container"
  >
    <div
      className="header__wrapper"
    >
      <div
        className="header__left"
      >
        <span
          className="header__logo-link header__logo-link--active"
        >
          <img
            alt="6 cities logo"
            className="header__logo"
            height="41"
            src="/img/logo.svg"
            width="81"
          />
        </span>
      </div>
      <nav
        className="header__nav"
      >
        <ul
          className="header__nav-list"
        >
          <li
            className="header__nav-item user"
          >
            <a
              className="header__nav-link header__nav-link--profile"
              href="/favorites"
              onClick={[Function]}
            >
              <div
                className="header__avatar-wrapper user__avatar-wrapper"
              >
                <img
                  alt="User avatar"
                  className="user__avatar"
                  src="https://htmlacademy-react-3.appspot.com/six-cities/pic.jpg"
                />
              </div>
              <span
                className="header__user-name user__name"
              >
                test@test.com
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</header>
`;

exports[`HeaderSnapshot should render Header with Link 1`] = `
<header
  className="header"
>
  <div
    className="container"
  >
    <div
      className="header__wrapper"
    >
      <div
        className="header__left"
      >
        <a
          className="header__logo-link"
          href="/"
          onClick={[Function]}
        >
          <img
            alt="6 cities logo"
            className="header__logo"
            height="41"
            src="/img/logo.svg"
            width="81"
          />
        </a>
      </div>
      <nav
        className="header__nav"
      >
        <ul
          className="header__nav-list"
        >
          <li
            className="header__nav-item user"
          >
            <a
              className="header__nav-link header__nav-link--profile"
              href="/favorites"
              onClick={[Function]}
            >
              <div
                className="header__avatar-wrapper user__avatar-wrapper"
              >
                <img
                  alt="User avatar"
                  className="user__avatar"
                  src="https://htmlacademy-react-3.appspot.com/six-cities/pic.jpg"
                />
              </div>
              <span
                className="header__user-name user__name"
              >
                test@test.com
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</header>
`;
